### ISSUE TITLE
[lsp-server] 🐞 LSP Crashes incase of error

### ISSUE THREAD

**Issue by tusharmath** [graphql/graphiql#3429]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

A graphql file with undefined types or directives crashes the server

### Expected Behavior

LSP should provide the error with swiggly lines. 

### Steps To Reproduce

1. Use the following files
    **f1.graphql**
    ```graphql
    schema @foo {
      query: Query
    }
    
    type Query {
      data: String
    }
    ```
    
    **f2.graphql**
    ```graphql
    directive @foo on SCHEMA
    ```
    
    **Config**
    ```yml
    schema:
      - ./f1.graphql
      - ./f2.graphql
    ```
2. Open in the project in VS Code and focus on the output's panel for GraphQL LSP.
3. Change the `foo` to `foos`.
4. Save the file.

**Crash Logs**
```
`))}var D2e=class{constructor(t){this._errors=[],this.schema=t}reportError(t,r){let n=Array.isArray(r)?r.filter(Boolean):r;this._errors.push(new Bnr.GraphQLError(t,{nodes:n}))}getErrors(){return this._errors}};function Vnr(e){let t=e.schema,r=t.getQueryType();if(!r)e.reportError("Query root type must be provided.",t.astNode);else if(!(0,cy.isObjectType)(r)){var n;e.reportError(`Query root type must be Object type, it cannot be ${(0,hb.inspect)(r)}.`,(n=w2e(t,A2e.OperationTypeNode.QUERY))!==null&&n!==void 0?n:r.astNode)}let i=t.getMutationType();if(i&&!(0,cy.isObjectType)(i)){var a;e.reportError(`Mutation root type must be Object type if provided, it cannot be ${(0,hb.inspect)(i)}.`,(a=w2e(t,A2e.OperationTypeNode.MUTATION))!==null&&a!==void 0?a:i.astNode)}let s=t.getSubscriptionType();if(s&&!(0,cy.isObjectType)(s)){var o;e.reportError(`Subscription root type must be Object type if provided, it cannot be ${(0,hb.inspect)(s)}.`,(o=w2e(t,A2e.OperationTypeNode.SUBSCRIPTION))!==null&&o!==void 0?o:s.astNode)}}function w2e(e,t){var r;return(r=[e.astNode,...e.extensionASTNodes].flatMap(n=>{var i;return(i=n?.operationTypes)!==null&&i!==void 0?i:[]}).find(n=>n.operation===t))===null||r===void 0?void 0:r.type}function Wnr(e){for(let r of e.schema.getDirectives()){if(!(0,Trt.isDirective)(r)){e.reportError(`Expected directive but got: ${(0,hb.inspect)(r)}.`,r?.astNode);continue}zM(e,r);for(let n of r.args)if(zM(e,n),(0,cy.isInputType)(n.type)||e.reportError(`The type of @${r.name}(${n.name}:) must be Input Type but got: ${(0,hb.inspect)(n.type)}.`,n.astNode),(0,cy.isRequiredArgument)(n)&&n.deprecationReason!=null){var t;e.reportError(`Required argument @${r.name}(${n.name}:) cannot be deprecated.`,[k2e(n.astNode),(t=n.astNode)===null||t===void 0?void 0:t.type])}}}function zM(e,t){t.name.startsWith("__")&&e.reportError(`Name "${t.name}" must not begin with "__", which is reserved by GraphQL introspection.`,t.astNode)}function znr(e){let t=Qnr(e),r=e.schema.getTypeMap();for(let n of Object.values(r)){if(!(0,cy.isNamedType)(n)){e.reportError(`Expected GraphQL named type but got: ${(0,hb.inspect)(n)}.`,n.astNode);continue}(0,$nr.isIntrospectionType)(n)||zM(e,n),(0,cy.isObjectType)(n)||(0,cy.isInterfaceType)(n)?(vrt(e,n),brt(e,n)):(0,cy.isUnionType)(n)?Jnr(e,n):(0,cy.isEnumType)(n)?Knr(e,n):(0,cy.isInputObjectType)(n)&&(Ynr(e,n),t(n))}}function vrt(e,t){let r=Object.values(t.getFields());r.length===0&&e.reportError(`Type ${t.name} must define one or more fields.`,[t.astNode,...t.extensionASTNodes]);for(let s of r){if(zM(e,s),!(0,cy.isOutputType)(s.type)){var n;e.reportError(`The type of ${t.name}.${s.name} must be Output Type but got: ${(0,hb.inspect)(s.type)}.`,(n=s.astNode)===null||n===void 0?void 0:n.type)}for(let o of s.args){let c=o.name;if(zM(e,o),!(0,cy.isInputType)(o.type)){var i;e.reportError(`The type of ${t.name}.${s.name}(${c}:) must be Input Type but got: ${(0,hb.inspect)(o.type)}.`,(i=o.astNode)===null||i===void 0?void 0:i.type)}if((0,cy.isRequiredArgument)(o)&&o.deprecationReason!=null){var a;e.reportError(`Required argument ${t.name}.${s.name}(${c}:) cannot be deprecated.`,[k2e(o.astNode),(a=o.astNode)===null||a===void 0?void 0:a.type])}}}}function brt(e,t){let r=Object.create(null);for(let n of t.getInterfaces()){if(!(0,cy.isInterfaceType)(n)){e.reportError(`Type ${(0,hb.inspect)(t)} must only implement Interface types, it cannot implement ${(0,hb.inspect)(n)}.`,ZH(t,n));continue}if(t===n){e.reportError(`Type ${t.name} cannot implement itself because it would create a circular reference.`,ZH(t,n));continue}if(r[n.name]){e.reportError(`Type ${t.name} can only implement ${n.name} once.`,ZH(t,n));continue}r[n.name]=!0,Gnr(e,t,n),Hnr(e,t,n)}}function Hnr(e,t,r){let n=t.getFields();for(let c of Object.values(r.getFields())){let f=c.name,l=n[f];if(!l){e.reportError(`Interface field ${r.name}.${f} expected but ${t.name} does not provide it.`,[c.astNode,t.astNode,...t.extensionASTNodes]);continue}if(!(0,yrt.isTypeSubTypeOf)(e.schema,l.type,c.type)){var i,a;e.reportError(`Interface field ${r.name}.${f} expects type ${(0,hb.inspect)(c.type)} but ${t.name}.${f} is type ${(0,hb.inspect)(l.type)}.`,[(i=c.astNode)===null||i===void 0?void 0:i.type,(a=l.astNode)===null||a===void 0?void 0:a.type])}for(let p of c.args){let h=p.name,g=l.args.find(y=>y.name===h);if(!g){e.reportError(`Interface field argument ${r.name}.${f}(${h}:) expected but ${t.name}.${f} does not provide it.`,[p.astNode,l.astNode]);continue}if(!(0,yrt.isEqualType)(p.type,g.type)){var s,o;e.reportError(`Interface field argument ${r.name}.${f}(${h}:) expects type ${(0,hb.inspect)(p.type)} but ${t.name}.${f}(${h}:) is type ${(0,hb.inspect)(g.type)}.`,[(s=p.astNode)===null||s===void 0?void 0:s.type,(o=g.astNode)===null||o===void 0?void 0:o.type])}}for(let p of l.args){let h=p.name;!c.args.find(y=>y.name===h)&&(0,cy.isRequiredArgument)(p)&&e.reportError(`Object field ${t.name}.${f} includes required argument ${h} that is missing from the Interface field ${r.name}.${f}.`,[p.astNode,c.astNode])}}}function Gnr(e,t,r){let n=t.getInterfaces();for(let i of r.getInterfaces())n.includes(i)||e.reportError(i===t?`Type ${t.name} cannot implement ${r.name} because it would create a circular reference.`:`Type ${t.name} must implement ${i.name} because it is implemented by ${r.name}.`,[...ZH(r,i),...ZH(t,r)])}function Jnr(e,t){let r=t.getTypes();r.length===0&&e.reportError(`Union type ${t.name} must define one or more member types.`,[t.astNode,...t.extensionASTNodes]);let n=Object.create(null);for(let i of r){if(n[i.name]){e.reportError(`Union type ${t.name} can only include type ${i.name} once.`,Srt(t,i.name));continue}n[i.name]=!0,(0,cy.isObjectType)(i)||e.reportError(`Union type ${t.name} can only include Object types, it cannot include ${(0,hb.inspect)(i)}.`,Srt(t,String(i)))}}function Knr(e,t){let r=t.getValues();r.length===0&&e.reportError(`Enum type ${t.name} must define one or more values.`,[t.astNode,...t.extensionASTNodes]);for(let n of r)zM(e,n)}function Ynr(e,t){let r=Object.values(t.getFields());r.length===0&&e.reportError(`Input Object type ${t.name} must define one or more fields.`,[t.astNode,...t.extensionASTNodes]);for(let a of r){if(zM(e,a),!(0,cy.isInputType)(a.type)){var n;e.reportError(`The type of ${t.name}.${a.name} must be Input Type but got: ${(0,hb.inspect)(a.type)}.`,(n=a.astNode)===null||n===void 0?void 0:n.type)}if((0,cy.isRequiredInputField)(a)&&a.deprecationReason!=null){var i;e.reportError(`Required input field ${t.name}.${a.name} cannot be deprecated.`,[k2e(a.astNode),(i=a.astNode)===null||i===void 0?void 0:i.type])}}}function Qnr(e){let t=Object.create(null),r=[],n=Object.create(null);return i;function i(a){if(t[a.name])return;t[a.name]=!0,n[a.name]=r.length;let s=Object.values(a.getFields());for(let o of s)if((0,cy.isNonNullType)(o.type)&&(0,cy.isInputObjectType)(o.type.ofType)){let c=o.type.ofType,f=n[c.name];if(r.push(o),f===void 0)i(c);else{let l=r.slice(f),p=l.map(h=>h.name).join(".");e.reportError(`Cannot reference Input Object "${c.name}" within itself through a series of non-null fields: "${p}".`,l.map(h=>h.astNode))}r.pop()}n[a.name]=void 0}}function ZH(e,t){let{astNode:r,extensionASTNodes:n}=e;return(r!=null?[r,...n]:n).flatMap(a=>{var s;return(s=a.interfaces)!==null&&s!==void 0?s:[]}).filter(a=>a.name.value===t.name)}function Srt(e,t){let{astNode:r,extensionASTNodes:n}=e;return(r!=null?[r,...n]:n).flatMap(a=>{var s;return(s=a.types)!==null&&s!==void 0?s:[]}).filter(a=>a.name.value===t)}function k2e(e){var t;return e==null||(t=e.directives)===null||t===void 0?void 0:t.find(r=>r.name.value===Trt.GraphQLDeprecatedDirective.name)}});var x2=rt(O2e=>{"use strict";Object.defineProperty(O2e,"__esModule",{value:!0});O2e.typeFromAST=N2e;var I2e=Ad(),xrt=Up();function N2e(e,t){switch(t.kind){case I2e.Kind.LIST_TYPE:{let r=N2e(e,t.type);return r&&new xrt.GraphQLList(r)}case I2e.Kind.NON_NULL_TYPE:{let r=N2e(e,t.type);return r&&new xrt.GraphQLNonNull(r)}case I2e.Kind.NAMED_TYPE:return e.getType(t.name.value)}}});var Iae=rt(tG=>{"use strict";Object.defineProperty(tG,"__esModule",{value:!0});t[Info  - 13:58:14] Connection to server got closed. Server will restart.
[Info  - 13:58:14] The graphql server has stopped running
[Info  - 13:58:14] The graphql server has stopped running
Debugger listening on ws://localhost:6009/ef66f750-9904-4959-b816-2b14938cbcd6
Debugger listening on ws://localhost:6009/ef66f750-9904-4959-b816-2b14938cbcd6
For help, see: https://nodejs.org/en/docs/inspector
[Info  - 13:58:14] {"type":"usage","messageType":"initialize"}
[Error - 13:58:14] WARNING: graphql-config error, only highlighting is enabled:
Unknown directive "@foos".
for more information on using 'graphql-config' with 'graphql-language-service-server', 
see the documentation at https://www.npmjs.com/package/graphql-language-service-server#user-content-graphql-configuration-file
[Error - 13:58:14] WARNING: graphql-config error, only highlighting is enabled:
Unknown directive "@foos".
for more information on using 'graphql-config' with 'graphql-language-service-server', 
see the documentation at https://www.npmjs.com/package/graphql-language-service-server#user-content-graphql-configuration-file
```

### Environment

* LSP Server Version:v0.8.19
* OS: Mac
* LSP Client: VSCode

**Comment by charleslai2000** [comment]
Encounter the same issue. I have vscode-graphql.cacheSchemaFileForLookup=true. Guess graphql-config can not validate the schema when schema has errors. The plugin just can not merge all schema into one schema cache file.

I suggest keep a backup of most recent usable cache schema and use it as input for LSP.

**Comment by thscott** [comment]
To add to this, it doesn't seem to crash until a graphql file containing errors is saved.

While modifying a graphql file, errors are indicated properly (and indication removed when fixed). If the file is saved while containing errors however, the graphql server crashes, and when it restarts, it no longer does anything except for syntax highlighting (logs `WARNING: graphql-config error, only highlighting is enabled`) -- so no error indications (even the existing error squiggles vanish), no jump to definition etc.

The only way to get it working properly again is to fix whatever errors there are in the graphql file, save the file, and then manually restart the graphql server (there is vs code command for this). Manually restarting the server while the file still contains errors also results in only syntax highlighting being enabled.

The `vscode-graphql.cacheSchemaFileForLookup` setting has no affect on this behaviour for me.

**Comment by tusharmath** [comment]
Happy to send a pr to fix if someone can tell me where to start.

**Comment by tusharmath** [comment]
I am actually surprised that it's been broken for so long, and doesn't seem to affect anyone. People either don't use it or have gotten used to the crash.

**Comment by acao** [comment]
huh, we just fixed a bug like this a few weeks ago. we started using graphql config loader to parse files again. a PR with tests that show this error would be welcome.

**Comment by A-N-uraag** [comment]
Hi @acao, I could reproduce this locally on `main`, so I submitted the fix instead of a test. Since it was just a wrap-around with try-catch, I did not add a test. Do let me know if it's mandatory to have one.

The issue:
While saving the files, the handler for `DidChangeWatchedFilesNotification` notification was being triggered, and this was not wrapped around with a try-catch in case of an error. I was not sure why this notification is triggered instead of `DidChangeTextDocumentNotification`, which has a try-catch already in it's handler.
